Add timeout argument to call_loop_with_reconnect_with_readonly

byronwolfman opened the pull request:
The redis-rb gem's `call_loop` def (seen here https://github.com/redis/redis-rb/commit/65f8886078e8b726dd78353edc99b5a5fd6a97c0#diff-597c124889a64c18744b52ef9687c572R130) accepts an optional `timeout` argument. Since it's optional, this gem does not necessarily need to submit timeout argument when it calls `call_loop_without_reconnect_with_readonly` (though doing so is polite).

The bigger issue is that when a redis subscription is created, that subscription then uses `call_loop` with the new timeout value. It submits two arguments where `call_loop_with_reconnect_with_readonly` only expects one and we stack trace with

    ArgumentError: wrong number of arguments (given 2, expected 1)
    /usr/local/bundle/gems/redis-reconnect_with_readonly-0.9.3/lib/redis/reconnect_with_readonly.rb:58:in `call_loop_with_reconnect_with_readonly'

This change adds the `timeout` argument to `call_loop_with_reconnect_with_readonly` which should fix the issue if I've done my research.

(Caveat: we've yet to meaningfully test this, but I'm opening up a PR now to at least get some additional eyes on it.)

byronwolfman committed:
Add timeout argument to call_loop_with_reconnect_with_readonly

The redis-rb gem's `call_loop` def (seen here https://github.com/redis/redis-rb/commit/65f8886078e8b726dd78353edc99b5a5fd6a97c0#diff-597c124889a64c18744b52ef9687c572R130) accepts an optional `timeout` argument. Since it's optional, this gem does not necessarily need to submit timeout argument when it calls `call_loop_without_reconnect_with_readonly` (though doing so is polite).

The bigger issue is that when a redis subscription is created, that subscription then uses `call_loop` with the new timeout value. It submits two arguments where `call_loop_with_reconnect_with_readonly` only expects one and we stack trace with

    ArgumentError: wrong number of arguments (given 2, expected 1)
    /usr/local/bundle/gems/redis-reconnect_with_readonly-0.9.3/lib/redis/reconnect_with_readonly.rb:58:in `call_loop_with_reconnect_with_readonly'

This change adds the `timeout` argument to `call_loop_with_reconnect_with_readonly` which should fix the issue if I've done my research.

byronwolfman commented:
As a non-rubyist I've naturally taken the shortest route to testing this by just wading into `~/.rbenv` and patching the living gem itself, which seems to have solved the underlying problem we were seeing.

sonots commented on lib/redis/reconnect_with_readonly.rb:
```diff
@@ -55,9 +55,9 @@ def call_with_reconnect_with_readonly(command, &block)
     end
 
-    def call_loop_with_reconnect_with_readonly(command, &block)
+    def call_loop_with_reconnect_with_readonly(command, timeout, &block)
```
Maybe, default value for timeout is needed?

byronwolfman replied:
Thanks for reviewing this! That's a good idea; since redis-rb's `call_loop` has a default timeout of 0, the replacement probably should too.

I'll commit an update.

byronwolfman committed:
Add default timeout of 0 to call_loop replacement

The redis-rb gem's `call_loop` sets a default timeout of 0, so `call_loop_with_reconnect_with_readonly` should mirror that behaviour and also set the same default timeout.

sonots commented on lib/redis/reconnect_with_readonly.rb:
```diff
@@ -55,9 +55,9 @@ def call_with_reconnect_with_readonly(command, &block)
+    def call_loop_with_reconnect_with_readonly(command, timeout = 0, &block)
       ReconnectWithReadonly.reconnect_with_readonly(self) do
-        call_loop_without_reconnect_with_readonly(command, &block)
+        call_loop_without_reconnect_with_readonly(command, timeout, &block)
```
Ah, `timeout` argument is available redis-rb >= 3.3.0.
So, this change makes redis-rb < 3.3.0 be broken.

Solution:

1) Support only redis-rb >= 3.3.0 by describing it on the gemspec file.
2) Branching by inspecting number of arguments of `call_loop_without_reconnect_with_readonly`. 

```
Redis::Client.instance_method(:call_loop_without_reconnect_with_readonly).parameters
```

Also, `arity` is available.

sonots replied:
redis-rb 3.3.0 is released on on 18 Apr 2016, it is long time ago.
So, I personally think describing version limitation on gemspec is fine.

Could you update gemspec?

byronwolfman replied:
Yep, done!

byronwolfman committed:
Declare dependency on redis >= 3.3.0

sonots commented:
Thanks!
